### ISSUE TITLE
C++ and Objective-C dependencies into C?

### ISSUE THREAD

**Issue by Lelelo1** [haxiomic/haxe-c-bridge#33]
Hey
I got great [instructions](https://community.haxe.org/t/missing-hxcpp-h-in-cpp-compilation/2975/20) in the Haxe community forums, and it worked out well for Haxe code to compile and be used in a watchOS project in swift (via an [Objective-C bridging header file](https://developer.apple.com/documentation/swift/imported_c_and_objective-c_apis/importing_objective-c_into_swift)).
I wonder though when it come to dependencies - is it restricted strictly C - or can Cpp be used like [LucenePlusPlus](https://github.com/luceneplusplus/LucenePlusPlus) in the watchOS projects as well?
My understanding so far about haxe in general is that the project becomes tied and dependent to for instance Java - if you use a Java dependency in Haxe.  But I am inexperienced in the realm of C, C++ and Objective-C, and HaxeCBridge for that matter.
So can I pick out C++ and Objective C dependencies, and have it compiled into C still?

**Comment by Lelelo1** [comment]
Is it possible to use a Objective-C library in haxe, output it in C++ and then used in a C# project, for instance? Or is that ability to use both C++ and Objective-c limited to the clock project? Thanks for samples!

**Comment by haxiomic** [comment]
Yes, it's possible to use Objective-C from haxe and then use the result from a C# project but you'll need to take care to understand how everything interconnects. Objective-C is largely syntax sugar over C++ if you rename your `.cpp` files to `.mm`, you can still compile them as C++ but now you can also write and interact with Objective-C. Haxe will do this for you if you add `-D objc`.

Once you've got that working and you've figured out externs so you can use the library you want in haxe, the next step is using the result from C#.

C# can call native code compiled into a dynamic library, which is like an exe but you don't run it, instead it contains native code for other programs to use (extension is `.dll` on windows, `.so` on linux and `.dylib` on macOS). You could use haxe-c-bridge for this step – it will give you a dynamic library and a header file with methods you've exposed from haxe. Here's some notes on calling native code from C#: https://stackoverflow.com/questions/11425202/is-it-possible-to-call-a-c-function-from-c-net 
The header from haxe-c-bridge should work out of the box with C#

**However**, it depends on what platform you're using – if it's on mac it'll work, linux and window probably not because I've not yet tested those platforms. The changes requires to haxe-c-bridge will be minor, I just don't have the setup and justification yet.

**Comment by Lelelo1** [comment]
My use case is to embed a search engine that can be used in a mobile app, watch projects, potentially existing and existing/upcoming [ARGlasses](https://community.haxe.org/t/heaps-3d-for-smart-glasses/3025).
I have always strived towards cross platform solutions. I am still learning and understanding the limits of haxe.
The mobile app is a Xamarin.Forms project, and I have read that embedding c++ libraries should be possible in the official docs. The watchOS project worked out with the dummy code as you already know
My ambition is to have this search engine and some custom (haxe) code available for as many platforms as possible.
